File: app/src/main/java/com/glucocontrol/data/repository/GlucoseRepositoryImpl.kt
package com.glucocontrol.data.repository

import com.glucocontrol.data.remote.dto.GlucoseReadingCreateDto
import com.glucocontrol.data.remote.dto.GlucoseReadingDto
import com.glucocontrol.data.remote.dto.GlucoseReadingUpdateDto
import com.glucocontrol.domain.model.GlucoseReading
import com.glucocontrol.domain.repository.GlucoseRepository
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.gotrue.auth
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flow
import java.time.LocalDate
import javax.inject.Inject

private const val TABLE = "glucose_readings"

/**
 * Implementación remota del repositorio de glucosa usando Supabase PostgREST.
 *
 * Cada Flow usa [refreshTrigger] con flatMapLatest para re-fetchear los datos
 * al producirse cualquier mutación (insert/update/delete). Así los ViewModels
 * reciben datos actualizados sin necesidad de navegar fuera y volver.
 *
 * Los flows envuelven la llamada de red con runCatching: si la red falla,
 * la tabla no existe o no hay sesión activa, emiten lista vacía (sin crash).
 */
@OptIn(ExperimentalCoroutinesApi::class)
class GlucoseRepositoryImpl @Inject constructor(private val supabase: SupabaseClient) : GlucoseRepository {

    // replay = 1 garantiza emisión inmediata al suscribirse; tryEmit nunca bloquea
    private val refreshTrigger = MutableSharedFlow<Unit>(replay = 1).also { it.tryEmit(Unit) }

    private fun invalidate() = refreshTrigger.tryEmit(Unit)

    private suspend fun currentUserId(): String = supabase.auth.currentUserOrNull()?.id
        ?: error("Sin sesión activa")

    override fun getReadingsByDate(date: LocalDate): Flow<List<GlucoseReading>> = refreshTrigger.flatMapLatest {
        flow {
            val result = runCatching {
                val userId = currentUserId()
                supabase.postgrest[TABLE]
                    .select {
                        filter {
                            eq("user_id", userId)
                            eq("date_epoch_day", date.toEpochDay())
                        }
                        order("time_seconds_of_day", Order.ASCENDING)
                    }
                    .decodeList<GlucoseReadingDto>()
                    .map { it.toDomain() }
            }
            emit(result.getOrDefault(emptyList()))
        }
    }

    override fun getReadingsByDateRange(from: LocalDate, to: LocalDate): Flow<List<GlucoseReading>> =
        refreshTrigger.flatMapLatest {
            flow {
                val result = runCatching {
                    val userId = currentUserId()
                    supabase.postgrest[TABLE]
                        .select {
                            filter {
                                eq("user_id", userId)
                                gte("date_epoch_day", from.toEpochDay())
                                lte("date_epoch_day", to.toEpochDay())
                            }
                            order("date_epoch_day", Order.ASCENDING)
                            order("time_seconds_of_day", Order.ASCENDING)
                        }
                        .decodeList<GlucoseReadingDto>()
                        .map { it.toDomain() }
                }
                emit(result.getOrDefault(emptyList()))
            }
        }

    override suspend fun insertReading(reading: GlucoseReading): Long {
        val userId = currentUserId()
        val dto = GlucoseReadingCreateDto.fromDomain(reading, userId)
        // postgrest-kt 2.x: insert ejecuta inmediatamente y retorna body vacío sin
        // Prefer:return=representation (propiedad privada en esta versión del SDK).
        supabase.postgrest[TABLE].insert(dto)
        invalidate()
        return 0L
    }

    override suspend fun updateReading(reading: GlucoseReading) {
        val userId = currentUserId()
        val dto = GlucoseReadingUpdateDto.fromDomain(reading, userId)
        supabase.postgrest[TABLE]
            .update(dto) {
                filter {
                    eq("id", reading.id)
                    eq("user_id", userId)
                }
            }
        invalidate()
    }

    override suspend fun deleteReading(id: Long) {
        val userId = currentUserId()
        supabase.postgrest[TABLE]
            .delete {
                filter {
                    eq("id", id)
                    eq("user_id", userId)
                }
            }
        invalidate()
    }

    override suspend fun getReadingById(id: Long): GlucoseReading? {
        val userId = currentUserId()
        return runCatching {
            supabase.postgrest[TABLE]
                .select {
                    filter {
                        eq("id", id)
                        eq("user_id", userId)
                    }
                }
                .decodeList<GlucoseReadingDto>()
                .firstOrNull()
                ?.toDomain()
        }.getOrNull()
    }
}

File: lib/data/repository/glucose_repository_impl.dart
import 'dart:async';

import 'package:glucocontrol/data/dto/glucose_reading_dto.dart';
import 'package:glucocontrol/data/remote/supabase_client_provider.dart';
import 'package:glucocontrol/domain/model/glucose_reading.dart';
import 'package:glucocontrol/domain/repository/glucose_repository.dart';

class GlucoseRepositoryImpl implements GlucoseRepository {
  // refreshTrigger invalida todos los Streams activos tras cualquier mutación
  final _refresh = StreamController<void>.broadcast();

  void _invalidate() => _refresh.add(null);

  String get _userId =>
      supabaseClient.auth.currentUser?.id ?? (throw Exception('No autenticado'));

  Stream<List<GlucoseReading>> _query(
    int fromEpochDay,
    int toEpochDay,
  ) async* {
    yield await _fetch(fromEpochDay, toEpochDay);
    await for (final _ in _refresh.stream) {
      yield await _fetch(fromEpochDay, toEpochDay);
    }
  }

  Future<List<GlucoseReading>> _fetch(int fromDay, int toDay) async {
    try {
      final rows = await supabaseClient
          .from(kTableGlucoseReadings)
          .select()
          .eq('user_id', _userId)
          .gte('date_epoch_day', fromDay)
          .lte('date_epoch_day', toDay)
          .order('date_epoch_day', ascending: true)
          .order('time_seconds_of_day', ascending: true);

      return (rows as List)
          .map((r) => GlucoseReadingDto.fromJson(r as Map<String, dynamic>).toDomain())
          .toList();
    } catch (_) {
      return [];
    }
  }

  // Usa año/mes/día tal cual (sin .toUtc()) — ver nota en DriftGlucoseRepository.
  int _toEpochDay(DateTime date) =>
      DateTime.utc(date.year, date.month, date.day).millisecondsSinceEpoch ~/
      86400000;

  @override
  Stream<List<GlucoseReading>> getReadingsByDate(DateTime date) {
    final day = _toEpochDay(date);
    return _query(day, day);
  }

  @override
  Stream<List<GlucoseReading>> getReadingsByDateRange(DateTime from, DateTime to) =>
      _query(_toEpochDay(from), _toEpochDay(to));

  @override
  Future<int> insertReading(GlucoseReading reading) async {
    final dto = GlucoseReadingCreateDto.fromDomain(reading, _userId);
    await supabaseClient.from(kTableGlucoseReadings).insert(dto.toJson());
    _invalidate();
    return 0; // Supabase sin Prefer:return=representation no devuelve el id
  }

  @override
  Future<void> updateReading(GlucoseReading reading) async {
    final dto = GlucoseReadingUpdateDto.fromDomain(reading, _userId);
    await supabaseClient
        .from(kTableGlucoseReadings)
        .update(dto.toJson())
        .eq('id', reading.id)
        .eq('user_id', _userId);
    _invalidate();
  }

  @override
  Future<void> deleteReading(int id) async {
    await supabaseClient
        .from(kTableGlucoseReadings)
        .delete()
        .eq('id', id)
        .eq('user_id', _userId);
    _invalidate();
  }

  @override
  Future<GlucoseReading?> getReadingById(int id) async {
    try {
      final rows = await supabaseClient
          .from(kTableGlucoseReadings)
          .select()
          .eq('id', id)
          .eq('user_id', _userId)
          .limit(1);

      final list = rows as List;
      if (list.isEmpty) return null;
      return GlucoseReadingDto.fromJson(list.first as Map<String, dynamic>)
          .toDomain();
    } catch (_) {
      return null;
    }
  }
}
